homeassistant/components/keba/__init__.py

```diff
@@ -0,0 +1,304 @@
+"""Support for KEBA charging stations."""
+import asyncio
+import json
+import logging
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_HOST
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'keba'
+SUPPORTED_COMPONENTS = ['binary_sensor', 'sensor', "lock"]
+
+SERVICE_UPDATE_STATE = 'update_state'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional('rfid', default='00845500'): cv.string,
+        vol.Optional('failsafe', default=False): cv.boolean,
+        vol.Optional('failsafe_timeout', default=30): cv.positive_int,
+        vol.Optional('failsafe_fallback', default=6): cv.positive_int,
+        vol.Optional('failsafe_save', default=0): cv.positive_int,
+        vol.Optional('refresh_interval', default=5): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+_SERVICE_MAP = {
+    'authorize ': 'async_authorize',
+    'deauthorize ': 'async_deauthorize',
+    'set_energy ': 'async_set_energy',
+    'set_curr': 'async_set_max_current',
+    'start': 'async_start',
+    'stop': 'async_stop'
+}
+
+
+async def async_setup(hass, config):
+    """Check connecitivty and version of KEBA charging station."""
+    # Setup KebaProtocol as communication handler
+    host = config[DOMAIN][CONF_HOST]
+    rfid = config[DOMAIN]['rfid']
+    refresh_interval = config[DOMAIN]['refresh_interval']
+
+    keba = KebaProtocol(hass, rfid, refresh_interval)
+    hass.data[DOMAIN] = keba
+
+    await hass.loop.create_datagram_endpoint(lambda: keba,
```

Comment:
Please extract the device specific interface code, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_component_code_review.html#4-communication-with-devices-services

Reply:
Thanks for the comment, I now extracted I/O code to pypi.

https://pypi.org/project/keba-kecontact/
https://github.com/dannerph/keba-kecontact